internal/qmp-gen/types.go

```diff
@@ -0,0 +1,532 @@
+				return fmt.Errorf("unexpected token %q (%T), wanted ']'", t, t)
+			}
+
+			field.Type = name(s)
+			field.List = true
+		default:
+			return fmt.Errorf("unexpected token %q (%T), wanted string or '['", t, t)
+		}
+
+		*fs = append(*fs, field)
+	}
+	return nil
+}
+
+type name string
+
+func (n name) Go() string {
+	if v, ok := builtinTypes[string(n)]; ok {
+		return v
+	}
+
+	return n.FieldName()
+}
+
+func (n name) FieldName() string {
+	out := []string{}
+	for _, p := range camelcase.Split(string(n)) {
+		if p == "-" || p == "_" {
+			continue
+		}
+		if upperWords[strings.ToLower(p)] {
+			p = strings.ToUpper(p)
+		} else {
+			p = strings.Title(strings.ToLower(p))
+		}
+		out = append(out, p)
+
+	}
+
+	return strings.Join(out, "")
+}
+
+func (n name) SimpleType() bool {
+	_, ok := builtinTypes[string(n)]
+	return ok
+}
+
+func (n name) InterfaceType(api map[name]interface{}) bool {
+	if n.SimpleType() {
+		return false
+	}
+	switch api[n].(type) {
+	case simpleUnion, flatUnion, alternate:
+		return true
+	}
+	return false
+}
+
+var builtinTypes = map[string]string{
+	"str":    "string",
+	"number": "float64",
+	"int":    "int64",
+	"int8":   "int8",
+	"int16":  "int16",
+	"int32":  "int32",
+	"int64":  "int64",
+	"uint8":  "uint8",
+	"uint16": "uint16",
+	"uint32": "uint32",
+	"uint64": "uint64",
+	"size":   "uint64",
+	"bool":   "bool",
+	"any":    "interface{}",
+}
+
+var upperWords = map[string]bool{
+	"cpu":   true,
+	"http":  true,
+	"https": true,
+	"id":    true,
+	"io":    true,
+	"ip":    true,
+	"json":  true,
+	"kvm":   true,
+	"nbd":   true,
+	"ok":    true,
+	"pci":   true,
+	"qmp":   true,
+	"ram":   true,
+	"tcp":   true,
+	"tls":   true,
+	"tpm":   true,
+	"ttl":   true,
+	"udp":   true,
+	"uuid":  true,
+	"vm":    true,
+	"vnc":   true,
+}
+
+// pyToJSON converts one QAPI type definition from its source
+// "json-ish python dict" representation to a string parseable by
+// encoding/json.
+func pyToJSON(py []byte) []byte {
+	var (
+		advanceTo byte
+		drop      bool
+	)
+	ret := make([]byte, 0, len(py))
+	for _, b := range py {
+		switch {
+		case advanceTo != 0:
+			// Inside a quoted string or a comment.
+			if b == advanceTo {
+				advanceTo = 0
+				drop = false
+			}
+		default:
+			switch b {
+			case '\'', '"':
+				// Start of a quoted string, advance to the end of it
+				// without looking for comments.
+				advanceTo = b
+			case '#':
+				// Start of a line comment, skip to EOL
+				advanceTo = '\n'
+				drop = true
+			}
+		}
+		if !drop {
+			if b == '\'' {
+				// In JSON, strings are always dquoted.
+				b = '"'
+			}
+			ret = append(ret, b)
+		}
+	}
+	return ret
+}
+
+func resolvePath(orig, new string) (string, error) {
+	u, err := url.Parse(orig)
+	if err != nil {
+		return "", err
+	}
+	u.Path = filepath.Join(filepath.Dir(u.Path), new)
+	return u.String(), nil
+}
+
+// getQAPI reads a QMP API spec file, from local disk or over HTTP(S).
+func getQAPI(path string) ([]byte, error) {
+	if path[0] == '/' {
```

Comment:
It may be a good idea to first `url.Parse()` here so we can handle relative filesystem paths.